rcl/include/rcl/context.h

```diff
@@ -0,0 +1,251 @@
+ * |  +------------v------------+
+ * |  |                         |
+ * |  | initialized but invalid +---> finalize all entities, then rcl_context_fini() +
+ * |  |                         |                                                    |
+ * |  +-------------------------+                                                    |
+ * |                                                                                 |
+ * +---------------------------------------------------------------------------------+
+ * ```
+ *
+ * A declared but not defined `rcl_context_t` instance is considered to be
+ * "uninitialized", and passing an uninitialized context to any functions will
+ * result in undefined behavior.
+ * Some functions, like `rcl_init()` require the context instance to be
+ * zero initialized (all members set to "zero" state) before use.
+ *
+ * Zero initialization of an `rcl_context_t` should be done with
+ * `rcl_get_zero_initialized_context()`, which ensures the context is in a safe
+ * state for initialization with `rcl_init()`.
+ *
+ * Initialization of an `rcl_context_t` should be done with `rcl_init()`, after
+ * which the context is considered both initialized and valid.
+ * After initialization it can be used in the creation of other entities like
+ * nodes and guard conditions.
+ *
+ * At any time the context can be invalidated by calling `rcl_shutdown()` on
+ * the `rcl_context_t`, after which the context is still initialized but now
+ * invalid.
+ *
+ * Invalidation indicates to other entities that the context was shutdown, but
+ * is still accessible for use during cleanup of themselves.
+ *
+ * After being invalidated, and after all of the entities which used it have
+ * been finalized, the context should be finalized with `rcl_context_fini()`.
+ *
+ * Finalizing the context while entities which have copies of it have not yet
+ * been finalized is undefined behavior.
+ * Therefore, the context's lifetime (between calls to `rcl_init()` and
+ * `rcl_context_fini()`) should exceed the lifetime of all entities which use
+ * it directly (e.g. nodes and guard conditions) or indirectly (e.g.
+ * subscriptions and topics).
+ */
+typedef struct rcl_context_t
+{
+  /// Global arguments for all nodes which share this context.
+  /** Typically generated by the parsing of argc/argv in `rcl_init()`. */
+  rcl_arguments_t global_arguments;
+  /// Implementation specific pointer.
+  struct rcl_context_impl_t * impl;
+  /// Private instance ID for this context.
+  /**
+   * A value of `0` indicates that it is either invalid or zero-initialized.
+   *
+   * Accessing the instance id should still be done using the function
+   * `rcl_context_get_instance_id()` because the instance id's type is an
+   * atomic and needs to be accessed properly to ensure safety.
+   *
+   * The instance id should not be changed manually - doing so is undefined
+   * behavior.
+   *
+   * The instance id cannot be protected within the `impl` pointer's type
```

Comment:
Could we instead update `rcl_context_get_instance_id()` to return `0` if the context is zero initialized? Then we could push the instance ID into the impl.

Reply:
The reason I wanted to avoid that was this case:

- thread 1: call rcl_context_get_instance_id()
- thread 1: if (context->impl)
- thread 2: rcl_context_fini(context) ... until complete
- thread 1: return context->impl->instance_id <--- nullptr access

Reply:
Okay, I see. Maybe the address of the impl should should double as the ID :stuck_out_tongue_winking_eye:

Reply:
That would be neat, but unfortunately once memory is free'd it can be reused (so a future init can get the same pointer as the current init if it is ever finalized).

Reply:
I pushed to address this by hiding it behind "storage" (i.e. uint8_t array), but it still cannot be put into `impl` due to the race condition: 9e7ce7a

Reply:
Also ros2/rcutils@fb42961.

Reply:
Well, at least people should be less tempted to modify a byte array than a single integer :smile: